### PR TITLE
test(plugin-openclaw): cover flush-plan ingest concurrency

### LINKED ISSUE #2393: test(plugin-openclaw): flush-plan ingest concurrency paths (rotation, recovery, lock loss) have no regression tests
## Tech-debt pass finding (Lane 5 weekly review, 2026-08-14)

**Category:** coverage gap · **Risk:** high

PR #2373 rewrote `packages/plugin-openclaw/src/delegate-flush-plan-ingest.ts` (+420/-71) with cross-process locking, atomic `.rotating`/`.inflight` file rotation, crash recovery, bounded reclaim passes, and ownership refresh. The PR's test additions all landed in `delegate-runtime.test.ts` (+386), and that file contains no coverage of flush-plan file rotation, `.inflight` recovery, or lock loss during posting (verified by grep: its lock/rotation tests are credential-rotation tests). No `delegate-flush-plan-ingest.test.ts` module exists in the tree.

These are exactly the paths a regression would silently corrupt: concurrent host append while a flush is posting, recovery from a crashed rotation, and continued appends after lock loss.

**Receipts:**
- `packages/plugin-openclaw/src/delegate-flush-plan-ingest.ts` (rewritten by #2373)
- `packages/plugin-openclaw/src/delegate-runtime.test.ts` — no `.rotating`/`.inflight`/flush-lock coverage
- Tree listing: no test module for the ingest file

**Suggested fix:** add a dedicated deterministic test module covering (a) concurrent append during an in-flight post, (b) recovery from `.rotating` and `.inflight` leftovers, (c) lock loss mid-flush declining rather than double-posting.

*Filed by the weekly tech-debt pass; ranked 1 of 7 this week.*

### PR DESCRIPTION
**Pull request by joshuaswarren**
## Summary

Add a dedicated deterministic test module for flush-plan ingestion. The tests cover host appends during posting, recovery from `.rotating` and `.inflight` leftovers, and lock loss before a second chunk.

Fixes #2393

## Type of change

- [ ] Bug fix
- [ ] Feature
- [ ] Docs
- [ ] Refactor
- [ ] Security / hardening

## Validation

- [ ] `npm run check-types` (root command unavailable because `pnpm` is not installed)
- [ ] `npm test`
- [ ] `npm run build`
- [ ] `bash scripts/check-review-patterns.sh`
- [x] Added/updated tests for behavior changes
- [ ] For retrieval/planner/cache/config changes: ran `docs/ops/pr-review-hardening-playbook.md`
- [ ] For high-risk memory/retrieval paths: ran `npm run test:entity-hardening`
- [ ] Ran `npm run review:cursor` locally (Cursor CLI unavailable)

Targeted validation passed:
- `npx tsx --test ./packages/plugin-openclaw/src/delegate-flush-plan-ingest.test.ts`
- `npm exec --yes pnpm@10.32.1 -- --filter @remnic/plugin-openclaw run check-types`

## Changelog

- [ ] Added/updated `CHANGELOG.md` under `## [Unreleased]`
- [x] Not needed (changeset added)

## Risk assessment

- [x] No secrets/tokens added
- [x] Backwards compatibility considered

## Notes for reviewers

The lock-loss test deliberately removes the lock while the first observe request is pending. It asserts that ingestion sends no second chunk and leaves the original snapshot for the replacement owner.

## PR workflow

- [x] PR scope is narrow
- [x] Synced with latest `main` before requesting AI review
- [x] Batched review fixes by subsystem instead of micro-pushing

---

> [!NOTE]
> **Low Risk**
> Test-only and changeset changes; no runtime or API behavior modified in this diff.
> 
> **Overview**
> Adds **`delegate-flush-plan-ingest.test.ts`** with integration-style tests against a local `/engram/v1/observe` stub, exercising **`ingestFlushPlanNotes`** without changing production code.
> 
> Coverage locks in three edge cases: **host appends while the first observe POST is in flight** (expects a second chunk and a clean plan), **recovery when `.inflight`, `.rotating`, and the live plan coexist** (expects inflight → rotating → plan ordering and cleanup), and **lock loss mid multi-chunk flush** (expects only one POST and the full snapshot left in `.inflight` for the next owner). A patch **changeset** documents the added regression coverage for `@remnic/plugin-openclaw`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 95c939123cb0f97901c5914ecebf2f5a9985944d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Patch release improves reliability for flush-plan ingestion during concurrent appends, recovery, and lock-loss scenarios.
  * Ensures recovered and in-progress files are processed in the correct order while preserving or cleaning up files appropriately.

* **Tests**
  * Added integration coverage for concurrent ingestion, recovery behavior, and interrupted flush operations.

### CODE REVIEW

File: packages/plugin-openclaw/src/delegate-flush-plan-ingest.test.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
+import assert from "node:assert/strict";
+import { appendFile, mkdir, mkdtemp, readFile, rm, unlink, writeFile } from "node:fs/promises";
+import http from "node:http";
+import os from "node:os";
+import path from "node:path";
+import test from "node:test";
+
+import type { DelegateDaemonTarget } from "./bridge.js";
+import { ingestFlushPlanNotes } from "./delegate-flush-plan-ingest.js";
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-openclaw/src/delegate-flush-plan-ingest.test.ts:9]: Declined. This is an internal source test for the module named in the issue. The package has no public export for this implementation, and adding one only to force a package-self import would expand the public API without testing a shipped contract. The relative imports match the neighboring tests and keep the change scoped to regression coverage.

File: packages/plugin-openclaw/src/delegate-flush-plan-ingest.test.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        };
+        await onObserve(body.messages?.[0]?.content ?? "");
+        response.statusCode = 200;
+        response.setHeader("content-type", "application/json");
+        response.end(JSON.stringify({ ok: true }));
+      })().catch(() => {
+        response.statusCode = 500;
+        response.end();
+      });
+    });
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-openclaw/src/delegate-flush-plan-ingest.test.ts:170]: Fixed in 95c939123. The HTTP stub now accepts only POST /engram/v1/observe with the expected bearer token and JSON content type. It also validates sessionKey and the single message content before invoking the test callback.
